### PR TITLE
Update repository methods to distinguish between jobs and pipelines

### PR DESCRIPTION
**Pull request by clairelin135**
Executing job CLI commands requires repositories to be able to differentiate between jobs and pipelines within. This PR updates `RepositoryData` and `CachingRepositoryData` to store and retrieve jobs and pipelines separately.

- All pipeline functions (e.g. `get_pipeline_names` and `has_pipeline`) will search among both jobs and pipelines
- All job functions will search only among jobs
- Enforces that all items added under the "jobs" key via `from_dict` repository instantiation are instances of `JobDefinition` or `GraphDefinition`

### CODE REVIEW

File: python_modules/dagster/dagster/core/definitions/repository.py [hunk 1]
```diff
@@ -381,10 +403,10 @@ def has_sensor(self, sensor_name):
 class CachingRepositoryData(RepositoryData):
     """Default implementation of RepositoryData used by the :py:func:`@repository <repository>` decorator."""
 
-    def __init__(self, pipelines, partition_sets, schedules, sensors):
+    def __init__(self, pipelines_and_jobs, partition_sets, schedules, sensors):
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/core/definitions/repository.py:406]: With my original comment on combining pipelines and jobs together, the intention was to still have distinct parameters for each in this function (and to have the cache continue to store them separately). The idea was more about the logical organization in the from_list function (where you have the benefit of knowing if an individual item is a JobDefinition or a PipelineDefinition, unlike when things are already in a dictionary).

File: python_modules/dagster/dagster/core/definitions/repository.py [hunk 2]
```diff
@@ -594,68 +621,129 @@ def from_list(cls, repository_definitions):
...
+        Args:
+            job_name (str): The name of the job.
+
+        Returns:
+            bool
+        """
+        check.str_param(job_name, "job_name")
+
+        if self._pipelines_and_jobs.has_definition(job_name):
+            if isinstance(self.get_pipeline(job_name), JobDefinition):
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/core/definitions/repository.py:682]: I think this is the main issue w/ keeping pipelines_and_jobs together under a single key. The old implementation did not have to instantiate the pipeline in order to tell if the pipeline existed or not.

Reverting back to having them in separate caches should fix this issue.

File: python_modules/dagster/dagster/core/definitions/repository.py [hunk 3]
```diff
@@ -594,68 +621,129 @@ def from_list(cls, repository_definitions):
...
+                            target_type=definition.target_type, target=definition.describe_target()
+                        )
+                    )
+                pipelines_and_jobs[coerced.name] = coerced
             else:
                 check.failed(f"Unexpected repository entry {definition}")
 
         return CachingRepositoryData(
-            pipelines=pipelines,
+            pipelines_and_jobs=pipelines_and_jobs,
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/core/definitions/repository.py:635]:
```
pipelines = # all of the PipelineDefinitions in this list
jobs = # all of the JobDefinitions in this list

return CachingRepositoryData(
            pipelines=pipelines,
            jobs=jobs,
            partition_sets=partition_sets,
            schedules=schedules,
            sensors=sensors,
        )
```

Then, get_all_pipelines() (and related functions) should return everything that's in either the pipelines cache or the jobs cache.

**Review by OwenKephart — CHANGES_REQUESTED** [review]

File: python_modules/dagster/dagster/core/definitions/repository.py (hunk 1 above)
Comment (clairelin135) [reply python_modules/dagster/dagster/core/definitions/repository.py]: Ahhhh ok this makes more sense. Reverted back to the original and changed the `from_list` function

File: examples/docs_snippets/docs_snippets/concepts/logging/python_logger.py [hunk 4]
```diff
@@ -23,10 +23,10 @@ def ambitious_op():
 
     # end_python_logger
     @graph
-    def thing():
+    def thing_one():
```
Comment (OwenKephart) [inline examples/docs_snippets/docs_snippets/concepts/logging/python_logger.py:26]: any reason for this change?

**Review by OwenKephart — APPROVED** [review]
minor comment about a potentially irrelevant change, but overall looks good!

File: examples/docs_snippets/docs_snippets/concepts/logging/python_logger.py (hunk 4 above)
Comment (clairelin135) [reply examples/docs_snippets/docs_snippets/concepts/logging/python_logger.py:26]: Adding two graphs with the same name to the same repository previously did not raise an error before this PR, even though it should because the graphs will be coerced to jobs. This was an edge case I caught and fixed in this PR, and this test had to be changed in order to not error afterwards.

Comment (OwenKephart) [reply examples/docs_snippets/docs_snippets/concepts/logging/python_logger.py:26]: ohhhh good point. that's great that we catch this now. 👍
